### PR TITLE
Merging XGBoost changes from 2.4

### PR DESCRIPTION
**Pull request by nvidianz**
### Description

Merged all the changes made to XGBoost in 2.4 back to main. This includes the user experience improvement and several bug fixes.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/app_opt/xgboost/histogram_based_v2/adaptors/grpc_client_adaptor.py [hunk 1]
```diff
@@ -99,6 +102,12 @@ def start(self, fl_ctx: FLContext):
         if not self.num_rounds:
             raise RuntimeError("cannot start - num_rounds is not set")
 
+        if not self.training_mode:
+            raise RuntimeError("cannot start - training_mode is not set")
+
+        if not self.xgb_params:
+            raise RuntimeError("cannot start - xgb_params is not set")
+
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/histogram_based_v2/adaptors/grpc_client_adaptor.py:110]: these checks should be moved into base "xgb_adaptor.py"?

File: nvflare/app_opt/xgboost/histogram_based_v2/adaptors/grpc_client_adaptor.py [hunk 2]
```diff
@@ -99,6 +102,12 @@ def start(self, fl_ctx: FLContext):
         if not self.num_rounds:
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/histogram_based_v2/adaptors/grpc_client_adaptor.py:102]: can "self.num_rounds" be set to "0"?
If not allowed,  should we change the error msg to something like:
```
raise RuntimeError("cannot start - num_rounds is not set or set to 0")
```

File: nvflare/app_opt/xgboost/histogram_based_v2/controller.py [hunk 3]
```diff
@@ -72,6 +75,9 @@ def __init__(
         Args:
             adaptor_component_id - the component ID of server target adaptor
             num_rounds - number of rounds
+            training_mode - Split mode (horizontal, vertical, horizontal_secure, vertical_secure)
+            xgb_params -The parameter passed to XGBoost
+            xgb_options - Other options for XGBoost
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/histogram_based_v2/controller.py:80]: what things belong to "xgb_params" vs xgb_options? what are their type?

File: nvflare/app_opt/xgboost/histogram_based_v2/fed_executor.py [hunk 4]
```diff
@@ -11,18 +11,19 @@
...
 from .executor import XGBExecutor
+from .sec.client_handler import ClientSecurityHandler
 
 
 class FedXGBHistogramExecutor(XGBExecutor):
     def __init__(
         self,
-        early_stopping_rounds,
-        xgb_params: dict,
         data_loader_id: str,
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/histogram_based_v2/fed_executor.py:27]: we could also have a follow up PR to update all the old xgboost examples to use the new executor/controller

File: nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py [hunk 5]
```diff
@@ -72,11 +71,15 @@ def initialize(self, fl_ctx: FLContext):
...
                 self.system_panic("writer should be type LogWriter", fl_ctx)
 
-    def _xgb_train(self, params: XGBoostParams, train_data: xgb.DMatrix, val_data) -> xgb.core.Booster:
+    def _xgb_train(self, num_rounds, xgb_params: dict, xgb_options: dict, train_data, val_data) -> xgb.core.Booster:
         """XGBoost training logic.
 
         Args:
-            params (XGBoostParams): xgboost parameters.
+            num_rounds: Number of rounds
+            xgb_params (XGBoostParams): xgboost parameters.
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py:79]: xgb_params (dict)

File: nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py [hunk 6]
```diff
@@ -92,14 +95,17 @@ def _xgb_train(self, params: XGBoostParams, train_data: xgb.DMatrix, val_data) -
         if flag and self._tb_dir:
             callbacks.append(TensorBoardCallback(self._tb_dir, tensorboard))
 
+        early_stopping_rounds = xgb_options.get("early_stopping_rounds", 0)
+        verbose_eval = xgb_options.get("verbose_eval", False)
+
         # Run training, all the features in training API is available.
         bst = xgb.train(
-            params.xgb_params,
+            xgb_params,
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py:103]: suggest update all docstring about this to "Booster params."

File: nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py [hunk 7]
```diff
@@ -72,11 +71,15 @@ def initialize(self, fl_ctx: FLContext):
             if not isinstance(self._metrics_writer, LogWriter):
                 self.system_panic("writer should be type LogWriter", fl_ctx)
 
-    def _xgb_train(self, params: XGBoostParams, train_data: xgb.DMatrix, val_data) -> xgb.core.Booster:
+    def _xgb_train(self, num_rounds, xgb_params: dict, xgb_options: dict, train_data, val_data) -> xgb.core.Booster:
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py:74]: for all the parameters to be passed into xgb.train method.
suggest either totally pack it as before a custom class call "XGBoostParams"
Or totally unpack it, so this method signature is similar to https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.train and docstring to explain each of the argument meaning.

Randomly group some into xgb_params, some standalone and some other into xgb_options is very confusing.

File: nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py [hunk 8]
```diff
@@ -109,40 +115,74 @@ def run(self, ctx: dict):
...
+            "xgboost_communicator": "federated",
             "federated_server_address": f"{self._server_addr}",
             "federated_world_size": self._world_size,
             "federated_rank": self._rank,
-            # FIXME: It should be possible to customize this or find a better location
-            # to distribut the shared object, preferably along side the nvflare Python
-            # package.
-            "federated_plugin": {"path": "/tmp/libproc_nvflare.so"},
         }
+
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py:143]: what is the correct way to work with XGBoost's latest code of secure xgboost?

File: nvflare/utils/cli_utils.py [hunk 9]
```diff
@@ -28,6 +28,14 @@ def get_home_dir() -> Path:
...
 
 
+def get_package_root() -> Path:
+    """
+    Get the nvflare package root folder, e.g.
+        /usr/local/python/3.10/lib/python3.10/site-packages/nvflare
+    """
+    return pathlib.Path(__file__).parent.parent.absolute().resolve()
+
+
```
Comment (YuanTingHsieh) [inline nvflare/utils/cli_utils.py:38]: there should be place that use this method?
should we include all those changes in another PR that does cherry-pick?

@chesterxgchen

**Review by YuanTingHsieh — CHANGES_REQUESTED** [review]
mostly good, some places we can enhance, can be this PR or follow up PR

File: nvflare/app_opt/xgboost/histogram_based_v2/adaptors/grpc_client_adaptor.py (hunk 1 above)
Comment (nvidianz) [reply nvflare/app_opt/xgboost/histogram_based_v2/adaptors/grpc_client_adaptor.py:110]: Those checks are redundant, already done in xgb_adaptor.py. Removed.

File: nvflare/app_opt/xgboost/histogram_based_v2/adaptors/grpc_client_adaptor.py (hunk 2 above)
Comment (nvidianz) [reply nvflare/app_opt/xgboost/histogram_based_v2/adaptors/grpc_client_adaptor.py:102]: This one is removed and added the > 0 check in xgb_adapter.py.

File: nvflare/app_opt/xgboost/histogram_based_v2/fed_executor.py (hunk 4 above)
Comment (nvidianz) [reply nvflare/app_opt/xgboost/histogram_based_v2/fed_executor.py:27]: Yes.

File: nvflare/app_opt/xgboost/histogram_based_v2/controller.py (hunk 3 above)
Comment (nvidianz) [reply nvflare/app_opt/xgboost/histogram_based_v2/controller.py:80]: Updated the document.

File: nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py (hunk 5 above)
Comment (nvidianz) [reply nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py:79]: Updated.

File: nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py (hunk 7 above)
Comment (nvidianz) [reply nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py:74]: The grouping is very clear. The params (Booster parameters) is passed as its own dictionary, everything else is packed into xgb_options. The params is specifically unpacked due to its importance.

File: nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py (hunk 8 above)
Comment (nvidianz) [reply nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py:143]: This is still coded against current API. Need another PR to work with XGBoost 2.2 API.

File: nvflare/utils/cli_utils.py (hunk 9 above)
Comment (nvidianz) [reply nvflare/utils/cli_utils.py:38]: This is used to find the .so in xgb_client_runner.py.

File: nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py (hunk 7 above)
Comment (YuanTingHsieh) [reply nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py:74]: Got it, then suggest either rename to "num_boost_round" or remove this "num_round" argument to be consistent

**Review by YuanTingHsieh — COMMENTED** [review]
Thanks, Just one comment.

File: nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py (hunk 7 above)
Comment (nvidianz) [reply nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py:74]: The num_round is a NVFlare convention. Most jobs have this parameter. So it's better to leave it alone.

Comment (YuanTingHsieh) [reply nvflare/app_opt/xgboost/histogram_based_v2/runners/xgb_client_runner.py:74]: Num rounds is ok on executor, but for xgb train method I think it is better to use same name as xgb.train.

But let's do that in follow up PRs.

**Review by YuanTingHsieh — APPROVED** [review]
LGTM; Enhancements can be done later as well.

**Pull request merged**